File: pnpm/crates/package-manager/src/remove_quarantine.rs
//! Strip the macOS Gatekeeper quarantine xattr from imported native binaries.
//!
//! Ported from pnpm v11's `fs/indexed-pkg-importer/src/removeQuarantine.ts`
//! and the `removeQuarantineFromNativeBinaries` sweep in
//! `fs/indexed-pkg-importer/src/index.ts`.
//!
//! macOS preserves extended attributes when files are copied, reflinked, or
//! hardlinked out of the content-addressable store. If a store blob carries
//! `com.apple.quarantine` (e.g. it was first written under a Gatekeeper-enabled
//! app such as a Git client), the quarantine propagates into `node_modules` and
//! Gatekeeper blocks the native binary from loading, even though pacquet has
//! already verified the file's integrity against the lockfile. Every caller of
//! [`import_indexed_dir`](crate::import_indexed_dir()) materializes from the
//! store, which is exactly pnpm's `resolvedFrom === 'store'` gate, so the sweep
//! runs after any populating import.

use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};

#[cfg(target_os = "macos")]
use std::{ffi::OsStr, process::Command};

#[cfg(target_os = "macos")]
const QUARANTINE_ATTR: &str = "com.apple.quarantine";

// Native binaries are the only files macOS Gatekeeper blocks for carrying a
// quarantine xattr; `.dll` is Windows-only and never relevant here.
#[cfg(target_os = "macos")]
const NATIVE_BINARY_EXTENSIONS: [&str; 3] = ["node", "dylib", "so"];

// Cap the bytes of file-path arguments per `xattr` call so a package with many
// (or very long) native-binary paths can't blow past the OS argv limit
// (ARG_MAX, ~1 MB on macOS). Well under the limit to leave room for argv0 and
// the environment block.
#[cfg(target_os = "macos")]
const MAX_ARG_BYTES: usize = 100_000;

/// Run a single batched quarantine sweep over the native binaries an import
/// just placed under `dir_path`. The relative entry paths come from the indexed
/// `cas_paths`, so the targets are `dir_path.join(entry)`. A no-op off macOS.
#[cfg(target_os = "macos")]
pub(crate) fn remove_quarantine_from_native_binaries(
    dir_path: &Path,
    cas_paths: &HashMap<String, PathBuf>,
) {
    let native_binaries: Vec<PathBuf> = cas_paths
        .keys()
        .filter(|entry| is_native_binary(entry))
        .map(|entry| dir_path.join(entry))
        .collect();
    remove_quarantine(&native_binaries);
}

#[cfg(not(target_os = "macos"))]
pub(crate) fn remove_quarantine_from_native_binaries(
    _dir_path: &Path,
    _cas_paths: &HashMap<String, PathBuf>,
) {
}

#[cfg(target_os = "macos")]
fn is_native_binary(entry: &str) -> bool {
    Path::new(entry).extension().and_then(OsStr::to_str).is_some_and(|ext| {
        NATIVE_BINARY_EXTENSIONS.iter().any(|known| known.eq_ignore_ascii_case(ext))
    })
}

/// Remove `com.apple.quarantine` from the given files, split into chunks that
/// stay under the OS argv limit. Paths are passed as separate arguments (never
/// interpolated into a shell), so package-controlled filenames cannot inject
/// shell commands. Non-fatal: errors that aren't a missing xattr or a missing
/// file are logged and the install continues.
#[cfg(target_os = "macos")]
fn remove_quarantine(file_paths: &[PathBuf]) {
    for chunk in chunk_by_arg_size(file_paths) {
        remove_quarantine_from_chunk(&chunk);
    }
}

#[cfg(target_os = "macos")]
fn remove_quarantine_from_chunk(file_paths: &[&Path]) {
    let output =
        Command::new("/usr/bin/xattr").arg("-d").arg(QUARANTINE_ATTR).args(file_paths).output();
    match output {
        Ok(output) if output.status.success() => {}
        Ok(output) => {
            // `xattr -d` exits non-zero when a file simply has no quarantine
            // xattr ("No such xattr"), the common case, and reports "No such
            // file" for entries the importer legitimately dropped or renamed.
            // Surface only errors that are not of those kinds.
            let stderr = String::from_utf8_lossy(&output.stderr);
            let real_errors: Vec<&str> = stderr
                .lines()
                .filter(|line| {
                    !line.trim().is_empty()
                        && !line.contains("No such xattr")
                        && !line.contains("No such file")
                })
                .collect();
            if !real_errors.is_empty() {
                tracing::warn!(
                    target: "pacquet::remove_quarantine",
                    errors = real_errors.join("\n"),
                    "failed to remove the macOS quarantine attribute",
                );
            }
        }
        Err(error) => {
            tracing::warn!(
                target: "pacquet::remove_quarantine",
                %error,
                "failed to spawn xattr to remove the macOS quarantine attribute",
            );
        }
    }
}

#[cfg(target_os = "macos")]
fn chunk_by_arg_size(file_paths: &[PathBuf]) -> Vec<Vec<&Path>> {
    let mut chunks: Vec<Vec<&Path>> = Vec::new();
    let mut chunk: Vec<&Path> = Vec::new();
    let mut chunk_bytes = 0;
    for file_path in file_paths {
        let bytes = file_path.as_os_str().len() + 1; // +1 for the argv null terminator
        if !chunk.is_empty() && chunk_bytes + bytes > MAX_ARG_BYTES {
            chunks.push(std::mem::take(&mut chunk));
            chunk_bytes = 0;
        }
        chunk.push(file_path.as_path());
        chunk_bytes += bytes;
    }
    if !chunk.is_empty() {
        chunks.push(chunk);
    }
    chunks
}

#[cfg(test)]
mod tests;

File: pnpm11/fs/indexed-pkg-importer/src/removeQuarantine.ts
import { execFileSync } from 'node:child_process'
import path from 'node:path'

import { globalWarn } from '@pnpm/logger'

const QUARANTINE_ATTR = 'com.apple.quarantine'
// Quarantine removal only ever runs on macOS, so the set lists the binary
// formats Gatekeeper guards there (.dll is Windows-only and never matches).
const NATIVE_BINARY_EXTENSIONS = new Set(['.node', '.dylib', '.so'])
// Cap the bytes of file-path arguments per `xattr` call so a package with many
// (or very long) native-binary paths can't blow past the OS argv limit
// (ARG_MAX, ~1 MB on macOS). Well under the limit to leave room for argv0 and
// the environment block.
const MAX_ARG_BYTES = 100_000

/**
 * Native binaries are the only files that macOS Gatekeeper blocks for carrying a
 * quarantine xattr, so removing it from anything else (JavaScript, text, etc.)
 * just wastes a syscall.
 */
export function isNativeBinary (filePath: string): boolean {
  return NATIVE_BINARY_EXTENSIONS.has(path.extname(filePath).toLowerCase())
}

/**
 * Remove the macOS Gatekeeper quarantine xattr (com.apple.quarantine) from the
 * given files using a single `xattr` invocation.
 *
 * macOS preserves extended attributes when pnpm copies or reflinks files out of
 * its content-addressable store. If a store blob carries com.apple.quarantine
 * (e.g. it was first written under a Gatekeeper-enabled app such as a Git
 * client), the quarantine propagates to node_modules and Gatekeeper blocks the
 * native binary from loading, even though pnpm already verified the file's
 * integrity against the lockfile. This mirrors Homebrew's behaviour of dropping
 * quarantine from verified downloads.
 *
 * File paths are passed as separate arguments rather than interpolated into a
 * shell command, so package-controlled filenames cannot inject shell commands.
 * They are split into chunks that stay under the OS argv limit.
 */
export function removeQuarantine (filePaths: string[]): void {
  if (process.platform !== 'darwin') return
  for (const chunk of chunkByArgSize(filePaths)) {
    removeQuarantineFromChunk(chunk)
  }
}

function removeQuarantineFromChunk (filePaths: string[]): void {
  try {
    execFileSync('/usr/bin/xattr', ['-d', QUARANTINE_ATTR, ...filePaths], {
      stdio: ['ignore', 'ignore', 'pipe'],
    })
  } catch (err: unknown) {
    // `xattr -d` exits non-zero when a file simply has no quarantine xattr to
    // remove ("No such xattr"), which is the common, expected case. It also
    // reports "No such file" for entries the importer legitimately dropped or
    // renamed (e.g. case-insensitive filename conflicts). Surface only errors
    // that are not of those kinds (e.g. permission denied).
    const realErrors = getStderr(err)
      .split('\n')
      .filter((line) => line.trim() !== '' && !line.includes('No such xattr') && !line.includes('No such file'))
    if (realErrors.length > 0) {
      globalWarn(`Failed to remove the macOS quarantine attribute:\n${realErrors.join('\n')}`)
    }
  }
}

function chunkByArgSize (filePaths: string[]): string[][] {
  const chunks: string[][] = []
  let chunk: string[] = []
  let chunkBytes = 0
  for (const filePath of filePaths) {
    const bytes = Buffer.byteLength(filePath) + 1 // +1 for the argv null terminator
    if (chunk.length > 0 && chunkBytes + bytes > MAX_ARG_BYTES) {
      chunks.push(chunk)
      chunk = []
      chunkBytes = 0
    }
    chunk.push(filePath)
    chunkBytes += bytes
  }
  if (chunk.length > 0) chunks.push(chunk)
  return chunks
}

function getStderr (err: unknown): string {
  if (typeof err === 'object' && err !== null && 'stderr' in err) {
    const stderr = (err as { stderr?: Buffer | string }).stderr
    if (stderr != null) return stderr.toString()
  }
  return err instanceof Error ? err.message : String(err)
}
